mempool/cat/pool_test.go

```diff
@@ -0,0 +1,669 @@
+	require.True(t, txExists("key1=0000=25"))
+	require.True(t, txExists("key8=0007=20"))
+	require.True(t, txExists("key10=0123456789abcdef=11"))
+	require.False(t, txExists("key3=0002=10"))
+	require.False(t, txExists("key9=0008=9"))
+	require.False(t, txExists("key7=0006=7"))
+
+	// Free up some space so we can add back previously evicted txs
+	err = txmp.Update(1, types.Txs{types.Tx("key10=0123456789abcdef=11")}, []*abci.ResponseDeliverTx{{Code: abci.CodeTypeOK}}, nil, nil)
+	require.NoError(t, err)
+	require.False(t, txExists("key10=0123456789abcdef=11"))
+	mustCheckTx(t, txmp, "key3=0002=10")
+	require.True(t, txExists("key3=0002=10"))
+
+	// remove a high priority tx and check if there is
+	// space for the previously evicted tx
+	require.NoError(t, txmp.RemoveTxByKey(types.Tx("key8=0007=20").Key()))
+	require.False(t, txExists("key8=0007=20"))
+	require.True(t, txmp.CanFitEvictedTx(types.Tx("key9=0008=9").Key()))
+}
+
+func TestTxPool_Flush(t *testing.T) {
+	txmp := setup(t, 0)
+	txs := checkTxs(t, txmp, 100, 0)
+	require.Equal(t, len(txs), txmp.Size())
+	require.Equal(t, int64(5690), txmp.SizeBytes())
+
+	rawTxs := make([]types.Tx, len(txs))
+	for i, tx := range txs {
+		rawTxs[i] = tx.tx
+	}
+
+	responses := make([]*abci.ResponseDeliverTx, len(rawTxs[:50]))
+	for i := 0; i < len(responses); i++ {
+		responses[i] = &abci.ResponseDeliverTx{Code: abci.CodeTypeOK}
+	}
+
+	txmp.Lock()
+	require.NoError(t, txmp.Update(1, rawTxs[:50], responses, nil, nil))
+	txmp.Unlock()
+
+	txmp.Flush()
+	require.Zero(t, txmp.Size())
+	require.Equal(t, int64(0), txmp.SizeBytes())
+}
+
+func TestTxPool_ReapMaxBytesMaxGas(t *testing.T) {
+	txmp := setup(t, 0)
+	tTxs := checkTxs(t, txmp, 100, 0) // all txs request 1 gas unit
+	require.Equal(t, len(tTxs), txmp.Size())
+	require.Equal(t, int64(5690), txmp.SizeBytes())
+
+	txMap := make(map[types.TxKey]testTx)
+	priorities := make([]int64, len(tTxs))
+	for i, tTx := range tTxs {
+		txMap[tTx.tx.Key()] = tTx
+		priorities[i] = tTx.priority
+	}
+
+	sort.Slice(priorities, func(i, j int) bool {
+		// sort by priority, i.e. decreasing order
+		return priorities[i] > priorities[j]
+	})
+
+	ensurePrioritized := func(reapedTxs types.Txs) {
+		reapedPriorities := make([]int64, len(reapedTxs))
+		for i, rTx := range reapedTxs {
+			reapedPriorities[i] = txMap[rTx.Key()].priority
+		}
+
+		require.Equal(t, priorities[:len(reapedPriorities)], reapedPriorities)
+	}
+
+	// reap by gas capacity only
+	reapedTxs := txmp.ReapMaxBytesMaxGas(-1, 50)
+	ensurePrioritized(reapedTxs)
+	require.Equal(t, len(tTxs), txmp.Size())
+	require.Equal(t, int64(5690), txmp.SizeBytes())
+	require.Len(t, reapedTxs, 50)
+
+	// reap by transaction bytes only
+	reapedTxs = txmp.ReapMaxBytesMaxGas(1200, -1)
+	ensurePrioritized(reapedTxs)
+	require.Equal(t, len(tTxs), txmp.Size())
+	require.Equal(t, int64(5690), txmp.SizeBytes())
+	require.GreaterOrEqual(t, len(reapedTxs), 16)
```

Comment:
No but when I played around I found that actually 20 transactions were returned. So better to assert that. I asked myself why not 21 because each tx is 57 bytes  so 57 * 21 = 1197 but I think 4 bytes are added to account for the proto encoding thus we get only 20.